### PR TITLE
UI to delete external services

### PR DESCRIPTION
**Pull request by nicksnyder**
Part of https://github.com/sourcegraph/sourcegraph/issues/914

screen recording 2018-12-04 at 12 53 pm

### CODE REVIEW

File: cmd/frontend/db/external_services.go [hunk 1]
```diff
@@ -81,6 +80,36 @@ func (c *externalServices) Update(ctx context.Context, id int64, update *Externa
 	})
 }
 
+type externalServiceNotFound struct {
```
Comment (beyang) [inline cmd/frontend/db/external_services.go]:
```suggestion
type externalServiceNotFoundErr struct {
```

It's customary to suffix error types of `Err` in our codebase (e.g., https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph@574ba50/-/blob/cmd/frontend/db/users.go#L32:4)

File: cmd/frontend/db/external_services.go [hunk 2]
```diff
@@ -81,6 +80,36 @@ func (c *externalServices) Update(ctx context.Context, id int64, update *Externa
...
+	return fmt.Sprintf("external service not found: %v", e.id)
+}
+
+func (e externalServiceNotFound) NotFound() bool {
+	return true
+}
+
+// Delete deletes an external service.
+//
+// 🚨 SECURITY: The caller must ensure that the actor is permitted to read external services.
```
Comment (beyang) [inline cmd/frontend/db/external_services.go]: Can you mention the explicit permission level here (looks like it's "admin"-level access)?

File: cmd/frontend/graphqlbackend/external_services.go [hunk 3]
```diff
@@ -85,6 +85,25 @@ func (*schemaResolver) UpdateExternalService(ctx context.Context, args *struct {
 	return &externalServiceResolver{externalService: externalService}, nil
 }
 
+func (*schemaResolver) DeleteExternalService(ctx context.Context, args *struct {
+	ExternalService graphql.ID
+}) (*EmptyResponse, error) {
+	// 🚨 SECURITY: Only site admins can delete orgs.
```
Comment (beyang) [inline cmd/frontend/graphqlbackend/external_services.go]:
```suggestion
	// 🚨 SECURITY: Only site admins can delete external services.
```

File: cmd/frontend/db/external_services.go (hunk 1 above)
Comment (slimsag) [reply cmd/frontend/db/external_services.go]: These should all be suffixed with `Error`: https://github.com/golang/go/wiki/Errors#naming

File: web/src/site-admin/SiteAdminExternalServicesPage.tsx [hunk 4]
```diff
@@ -30,12 +40,63 @@ class ExternalServiceNode extends React.PureComponent<ExternalServiceNodeProps,
...
+                    if (this.props.onDidUpdate) {
+                        this.props.onDidUpdate()
+                    }
+                },
+                err => this.setState({ loading: false, errorDescription: err.message })
+            )
+    }
+}
+
+export function deleteExternalService(externalService: GQL.ID): Observable<void> {
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminExternalServicesPage.tsx]: Doesn't need to be exported I think

File: web/src/site-admin/SiteAdminExternalServicesPage.tsx [hunk 5]
```diff
@@ -1,23 +1,33 @@
...
 interface ExternalServiceNodeProps {
     node: GQL.IExternalService
-    onDidUpdate?: () => void
+    onDidUpdate: () => void
 }
 
-class ExternalServiceNode extends React.PureComponent<ExternalServiceNodeProps, {}> {
+interface ExternalServiceNodeState {
+    loading: boolean
+    errorDescription?: string
```
Comment (felixfbecker) [inline web/src/site-admin/SiteAdminExternalServicesPage.tsx]: Is this displayed anywhere?

**Pull request merged**